Add particle property output

gassmoeller opened the pull request:
This is a preliminary implementation of the particle property output. Currently it only allows output of all properties (so no selection of properties) as scalars (so no vector or tensors), but I wanted to open the PR to let @blaisb and @luca-heltai have an early look. Will extend it to vector and tensor properties later today. Do you agree to put the information as optional arguments into `build_patches` or should we take a different route (e.g. an `add_particle_information` function)?

bangerth commented on source/particles/data_out.cc:
```diff
@@ -23,10 +23,41 @@ namespace Particles
+
+    for (const auto &data_component_name : data_component_names)
+      {
+        dataset_names.emplace_back(data_component_name);
```
Or this:
```
  dataset_names.insert (dataset_names.end(), data_component_names.begin(), data_component_names.end());
```

luca-heltai commented:
Doxygen is not happy about missing documentation:

~~~

0.00s$ cat doxygen.log && ! [ -s doxygen.log ]
/home/travis/build/dealii/dealii/include/deal.II/particles/data_out.h:71: warning: The following parameters of Particles::DataOut::build_patches(const Particles::ParticleHandler< dim, spacedim > &particles, const std::vector< std::string > &names=std::vector< std::string >(), const std::vector< DataComponentInterpretation::DataComponentInterpretation > &data_component_interpretation=std::vector< DataComponentInterpretation::DataComponentInterpretation >()) are not documented:
  parameter 'names'
  parameter 'data_component_interpretation'
The command "cat doxygen.log && ! [ -s doxygen.log ]" exited with 1.
cache.2
store build cache
~~~

gassmoeller commented:
I adressed the comments and added documentation.

The missing piece was the output of vector and tensor quantities, which I added now. In the process I found a bug in DataOutBase that only appears for tensor output if dim < spacedim. Just to be sure, in this case we would expect output tensors to be of type `Tensor<rank,spacedim>` is that correct? Because if not, particle tensor properties would be `Tensor<rank,0>`, which seems wrong.

I have one remaining question: Right now by default we do not output any particle properties, only if names and interpretations are handed over we output all of the properties. Would it make sense to assume the default should be to always output all properties as scalars (with made up names like `property_1`)? It would seem reasonable to me, but on the other hand I know of at least one ASPECT user who attaches thousands of properties to particles and does not want to output them, so maybe the current default (do not output anything) is more sensible? We could always add an additional argument later that allows to individually choose which properties to output (e.g. a std::vector<bool> of length n_properties).

blaisb commented:
> I adressed the comments and added documentation.
> 
> The missing piece was the output of vector and tensor quantities, which I added now. In the process I found a bug in DataOutBase that only appears for tensor output if dim < spacedim. Just to be sure, in this case we would expect output tensors to be of type `Tensor<rank,spacedim>` is that correct? Because if not, particle tensor properties would be `Tensor<rank,0>`, which seems wrong.
> 
> I have one remaining question: Right now by default we do not output any particle properties, only if names and interpretations are handed over we output all of the properties. Would it make sense to assume the default should be to always output all properties as scalars (with made up names like `property_1`)? It would seem reasonable to me, but on the other hand I know of at least one ASPECT user who attaches thousands of properties to particles and does not want to output them, so maybe the current default (do not output anything) is more sensible? We could always add an additional argument later that allows to individually choose which properties to output (e.g. a std::vector of length n_properties).

That is a very good questions. I think it would make sense that the default behavior would be to output all properties as scalar, but adding the option to disable properties is a very good idea!

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -5421,7 +5421,7 @@ namespace DataOutBase
                       {
                         const auto ind =
-                          Tensor<2, dim>::unrolled_to_component_indices(c);
+                          Tensor<2, spacedim>::unrolled_to_component_indices(c);
```
I suppose this matches what we do for vectors?

gassmoeller replied:
Actually no. I think I can write this better, please wait with merging, I will update the PR.

bangerth commented on source/particles/data_out.cc:
```diff
@@ -59,6 +111,140 @@ namespace Particles
+               unsigned int,
+               std::string,
+               DataComponentInterpretation::DataComponentInterpretation>>
+  DataOut<dim, spacedim>::get_nonscalar_data_ranges() const
```
I hope that this is a 1:1 copy of the `DataOut_DoFData` function, right? Possibly shifting everything by one for the id.

gassmoeller replied:
This is a 1:1 copy in which I removed all the code for complex output properties (because particles can not carry complex properties, unlike the DoF data).

bangerth replied:
Ah, nice. Yes, that would have made things simpler...

bangerth commented on source/particles/data_out.cc:
```diff
@@ -16,17 +16,57 @@
+    data_component_interpretations.insert(
+      data_component_interpretations.end(),
+      data_component_interpretations_.begin(),
+      data_component_interpretations_.end());
```
Hm, I guess my suggestion didn't actually make it more readable...

But at least it's faster this way! ;-)

bangerth commented:
OK to merge if it passes the tests, though I'm curious about the questions I have.

bangerth commented:
/rebuild

gassmoeller commented:
I updated the patch to `DataOutBase` in the last commit, which now unifies the handling of vectors and tensors inside `DataOutBase::write_vtu_main` and additionally simplifies the code. Ready from my side if you agree and the CI is happy.

gassmoeller commented:
Wait a minute, I think this is wrong (wrong order of components when converting tensors). Will get back to it.

gassmoeller commented:
Ok, I fixed it. The DataOutBase should determine the correct dimension based on the handed over output data, not its template parameters. This is what we do for vectors. So this is now ready to go from my side.

bangerth commented on source/base/data_out_base.cc:
```diff
@@ -5415,13 +5415,24 @@ namespace DataOutBase
-                          Tensor<2, dim>::unrolled_to_component_indices(c);
+                          Tensor<2, 2>::unrolled_to_component_indices(c);
+                        vtk_data[ind[0]][ind[1]] =
+                          data_vectors(first_component + c, n);
```
Yes, this looks right. You're just going to populate the top 2x2 sub-block.

I don't think the previous code was wrong, though. It relied on the fact that size==4 in 2d and size==9 whenever we got an output patch in 3d. But this is cleaner.

bangerth commented:
/rebuild
